### PR TITLE
DAOS-11605 common: basic ad tx undo/redo

### PR DESCRIPTION
**Pull request by liuxuezhao**
Add basic implementation for umem tx undo/redo.
Required-githooks: true

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

File: src/common/ad_tx.c [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+	D_ALLOC_ACT(it_redo, UMEM_ACT_MOVE, size);
+	if (it_redo == NULL) {
+		D_FREE(it_undo);
+		return -DER_NOMEM;
+	}
+
+	memcpy(it_undo->it_act.ac_copy.payload, dst, size);
+	it_undo->it_act.ac_copy.addr = ad_ptr2addr(dst);
+	it_undo->it_act.ac_copy.size = size;
+	AD_TX_ACT_ADD(tx, it_undo, ACT_UNDO);
```
Comment (gnailzenh) [inline src/common/ad_tx.c:375]: I think we probably should copy the whole region:
 addr = min(dst, src)
 end = max(dst + size, src + size)
 ac_copy.addr = addr;
 ac_copy.size = (end - addr);
 memcpy(....payload, addr, (end -addr))

Comment (liuxuezhao) [reply src/common/ad_tx.c:375]: OK, I see your point, if the dst and src buf overlapped, better to protect it by you suggested method.

But if the dst and src buf don't overlap, seems we only need to protect dst buf here? (user can explicitly protect src buf if needed).

File: src/common/ad_tx.c [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+			D_INIT_LIST_HEAD(&it->it_link);					\
+			it->it_act.ac_opc = opc;					\
+		}									\
+	} while (0)
+
+/** start a ad-hoc memory transaction */
+int
+ad_tx_begin(struct ad_blob *blob, struct ad_tx *tx)
+{
+	static __thread uint64_t	tx_id;
```
Comment (NiuYawei) [inline src/common/ad_tx.c:46]: What is this per-thread tx_id for?

File: src/common/ad_tx.c [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+		else if (opc == UMEM_ACT_CSUM)						\
+			D_ALLOC(it, offsetof(struct umem_act_item,			\
+					     it_act.ac_csum.csum[size]));		\
+		else									\
+			D_ALLOC_PTR(it);						\
+		if (likely(it != NULL)) {						\
+			D_INIT_LIST_HEAD(&it->it_link);					\
+			it->it_act.ac_opc = opc;					\
+		}									\
+	} while (0)
```
Comment (NiuYawei) [inline src/common/ad_tx.c:40]: I'm bit worried about the memory allocations, now basically every memory operation (over tmpfs) will cause a memory allocation, shouldn't we just allocate a vector to put all 'actions' in (the vector could grow on demand)? That could save lots of small allocations.

File: src/common/ad_tx.c (hunk 2 above)
Comment (liuxuezhao) [reply src/common/ad_tx.c:46]: just to temporarily generate the tx_id, I'm not sure what kind of tx_id should use later, currently it is just a placeholder can can be change later when needed. Seems the tx_id is not really needed/useful, we can remove or change later.

File: src/common/ad_tx.c (hunk 3 above)
Comment (liuxuezhao) [reply src/common/ad_tx.c:40]: the action allocation is using standard dram allocator, it seems commonly already with thread-cache slab in the allocator.
with our own slab vector for actions can be considered to add in following patch, although I am not sure if can get big benefit.

File: src/common/ad_tx.c (hunk 1 above)
Comment (liuxuezhao) [reply src/common/ad_tx.c:330]: as discussed, the original code is fine for this.

**Review by NiuYawei — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

**Review by gnailzenh — APPROVED** [review]

**Pull request merged**